rocketpy/motors/TankGeometry.py

```diff
@@ -0,0 +1,376 @@
+
+try:
+    from functools import cached_property
+except ImportError:
+    from rocketpy.tools import cached_property
+
+
+class TankGeometry:
+    """Class to define the geometry of a tank. It is used to calculate the
+    geometrical properties such as volume, area and radius. The tank is
+    axi-symmetric, and its geometry is defined by a set of Functions that are
+    used to calculate the radius as a function of height.
+    """
+
+    def __init__(self, geometry_dict=dict()):
+        """Initialize TankGeometry class.
+
+        Parameters
+        ----------
+        geometry_dict : dict, optional
+            Dictionary containing the geometry of the tank. The geometry is
+            calculated by a PiecewiseFunction. Hence, the dict keys are disjoint
+            tuples containing the lower and upper bounds of the domain of the
+            corresponding Function, while the values correspond to the radius
+            function from a axis of symmetry.
+        """
+        self.geometry = geometry_dict
+        return None
+
+    @property
+    def geometry(self):
+        """
+        The dictionary containing the geometry of the tank.
+
+        Returns
+        -------
+        dict
+            Dictionary containing the geometry of the tank.
+        """
+        return self._geometry
+
+    @geometry.setter
+    def geometry(self, geometry_dict):
+        """
+        Sets the geometry of the tank.
+
+        Parameters
+        ----------
+        geometry_dict : dict
+            Dictionary containing the geometry of the tank.
+        """
+        self._geometry = dict()
+        for domain, function in geometry_dict.items():
+            self.add_geometry(domain, function)
+
+    @funcify_method("Height (m)", "radius (m)", extrapolation="zero")
+    def radius(self):
+        """
+        The radius of the tank as a function of height.
+
+        Returns
+        -------
+        rocketpy.Function
+            Piecewise defined Function of tank radius.
+        """
+        return self.radius
+
+    @cached_property
+    def average_radius(self):
+        """
+        The average radius of the tank.
+
+        Returns
+        -------
+        float
+            Average radius of the tank.
+        """
+        return self.radius.average(self.bottom, self.top)
+
+    @property
+    def bottom(self):
+        """
+        The bottom of the tank. It is the lowest coordinate that belongs to
+        the domain of the geometry.
+
+        Returns
+        -------
+        float
+            Bottom coordinate of the tank.
+        """
+        return min(self._geometry.keys())[0]
+
+    @property
+    def top(self):
+        """
+        The top of the tank. It is the highest coordinate that belongs to
+        the domain of the geometry.
+
+        Returns
+        -------
+        float
+            Top coordinate of the tank.
+        """
+        return max(self._geometry.keys())[1]
+
+    @cached_property
```

Comment:
this calculation is so simple, why cache it?

Reply:
That was a mistake, fixed on 80daa46310121108ea6557f44e5c4663c2892ef3.